services/reddit/reddit-base.js

```diff
@@ -0,0 +1,91 @@
+import Joi from 'joi'
+import { BaseJsonService } from '../index.js'
+
+const tokenSchema = Joi.object({
+  access_token: Joi.string().required(),
+  expires_in: Joi.number(),
+  scope: Joi.string(),
+  token_type: Joi.string(),
+})
+
+// Abstract class for Reddit badges
+// Authorization flow based on https://github.com/reddit-archive/reddit/wiki/OAuth2#application-only-oauth.
+export default class RedditBase extends BaseJsonService {
```

Comment:
This is partly based on https://github.com/badges/shields/blob/master/services/twitch/twitch-base.js. However:
* there are some annoying differences compared to Twitch's OAuth flow.
* the logic in the Twitch code is partly broken. For example, the loop [here](https://github.com/badges/shields/blob/1b00489f577d28595fe28c2316b3336f512f500f/services/twitch/twitch-base.js#L87) does not refresh the request object, so we're querying new access tokens but still using the old one in our retries, which defeats the whole purpose of the retry loop. Also, I noticed whilst experimenting that err.response could be undefined in some cases, making the conditional blow up.

I decided to just try to make things work for Reddit as a fist iteration, and will came back to rework things at a later point.

Reply:
Yeah agreed. Last time I looked at the code for twitch I thought there are some issues that want fixing. Lets not copy the approach.